### PR TITLE
Turning on preqxx standalone tests on sandiatoss

### PR DESCRIPTION
**Pull request by oksanaguba**
Adds homme preqxx standalone tests with bfb comparison to F to sandiatoss3. Other baseline machines that are affected are compy and melvin. Their *-bfb.cmake files will be added later; so HOMMEBFB suite will be broken. 

Reducing number of preqxx tests to 3, a few tweaks in homme cmake.

[BFB] but will need new HOMMEBFB standalone baselines.

### CODE REVIEW

**Comment by oksanaguba** [comment]
Time limit does not need to be adjusted. C vs F tests fail due to issue with jw_baro test init, will be fixed in pr #3094. However, unit tests fail, too:
```

      Start 58: infrastructure_ut_test
Could not find executable infrastructure_ut
Looked in the following places:
infrastructure_ut
infrastructure_ut
Release/infrastructure_ut
Release/infrastructure_ut
Debug/infrastructure_ut
Debug/infrastructure_ut
MinSizeRel/infrastructure_ut
MinSizeRel/infrastructure_ut
RelWithDebInfo/infrastructure_ut
RelWithDebInfo/infrastructure_ut
Deployment/infrastructure_ut
Deployment/infrastructure_ut
Development/infrastructure_ut
Development/infrastructure_ut
Unable to find executable: infrastructure_ut
58/63 Test #58: infrastructure_ut_test ......................................***Not Run   0.00 sec
      Start 59: random_init_ut_test
Could not find executable random_init_ut
....

         58 - infrastructure_ut_test (Not Run)
         59 - random_init_ut_test (Not Run)
         60 - boundary_exchange_ut_test (Failed)
         61 - sphere_op_ut_test (Failed)
         62 - limiters_ut_test (Failed)
         63 - remap_ut_test (Failed)
```

**Comment by oksanaguba** [comment]
Other machines to consider for these tests are melvin and anvil. Melvin seems to not to be tuned for performance (layout, binding, environment?) and its tests take very long. For anvil, there are some issues with a gcc compiler.

**Comment by jgfouca** [comment]
@oksanaguba , since this is WIP, I assume you don't want me to merge it to next yet?

**Comment by oksanaguba** [comment]
@jgfouca not yet, thanks. i am not sure why unit tests fail now, from vaguely remembering, there could have been an issue with remap test, but not the rest. also, i wonder whether anyone would object turning on preqxx and fpmodel=strict on skybridge. @mt5555  would this be ok with you?

**Comment by rljacob** [comment]
@oksanaguba what is the status of this PR?

**Comment by oksanaguba** [comment]
Still in works, i did not have time to look at this.

**Comment by rljacob** [comment]
@oksanaguba can you finish this before opening more WIP PRs?

**Comment by oksanaguba** [comment]
Can we have 2 WIP PRs per person as a limit?

**Comment by rljacob** [comment]
You can have all you want in your fork.

**Comment by oksanaguba** [comment]
I want a wip PR to keep conversation. If I open a PR from my fork branch to my fork master and keep conversation there 1) people won't see what's going on 2) if I want to redefine a base branch later, I cannot redefine base branch from a different fork https://github.com/isaacs/github/issues/805 . So, I can create a new PR and the conversation is lost.

**Comment by oksanaguba** [comment]
Now I don't have issues with unit tests:
```
[onguba@skybridge-login9 bxx2]$ ctest -L unit
Test project /ascldap/users/onguba/runhomme/bxx2
    Start 58: infrastructure_ut_test
1/6 Test #58: infrastructure_ut_test ...........   Passed    0.52 sec
    Start 59: random_init_ut_test
2/6 Test #59: random_init_ut_test ..............   Passed    0.33 sec
    Start 60: boundary_exchange_ut_test
3/6 Test #60: boundary_exchange_ut_test ........   Passed    0.37 sec
    Start 61: sphere_op_ut_test
4/6 Test #61: sphere_op_ut_test ................   Passed    0.35 sec
    Start 62: limiters_ut_test
5/6 Test #62: limiters_ut_test .................   Passed    0.32 sec
    Start 63: remap_ut_test
6/6 Test #63: remap_ut_test ....................   Passed    0.30 sec
```

File: components/homme/cmake/machineFiles/sandiatoss3-bfb.cmake [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+# Openmpi 1.8 only
+SET (USE_MPI_OPTIONS " --map-by node:SPAN " CACHE FILEPATH "")
+
+# Openmpi 1.6
+#SET (USE_MPI_OPTIONS "-loadbalance" CACHE FILEPATH "")
+
+#turn on preqxx target and thus strict fpmodel for F vs CXX comparison
+SET (ADD_Fortran_FLAGS "-traceback -fp-model strict -qopenmp -O1" CACHE STRING "")
+SET (ADD_C_FLAGS "-traceback -fp-model strict -qopenmp -O3" CACHE STRING "")
+SET (ADD_CXX_FLAGS "-traceback -fp-model strict -qopenmp -O3" CACHE STRING "")
```
Comment (ambrad) [inline components/homme/cmake/machineFiles/sandiatoss3-bfb.cmake]: O1 vs O3 are from previous experience, right?

Comment (oksanaguba) [reply components/homme/cmake/machineFiles/sandiatoss3-bfb.cmake]: Yes, I wasn't even thinking -- these are settings that worked on skybridge before. They still work, all cxx_vs_f90 tests (profile dev) pass.

**Comment by oksanaguba** [comment]
Still testing the last change: getting
```
No space left on device
FATAL ERROR: fwrite on file failed
``` 
on skybridge.

In more detail, it seems that during model build tests in test_exec that are used for baselines are built, but unit tests did not. When model runs, it completes all 'baseline' runs and then tries to build unit tests, which are built partially till the error is hit. I can try to mark unit tests as baseline tests as well. They are probably not built because if no -g or -c is given, then homme.py tries to do make baselines and then make check.

**Comment by oksanaguba** [comment]
For last commit cime run passed. it is my assumption that now if unit tests (builds or runs) are broken, it would be caught in either cime or standalone homme for any combination of runs (make check, make baseline, etc).

**Comment by oksanaguba** [comment]
New classes seem to work. I tried to break a homme class by not supplying cmakesuffix in hommebaseclass and build got broken, but the error message that is in the base class did not show up. Last thing to do is to reduce # of cxx preqx tests, so that we can do more theta tests.

**Comment by oksanaguba** [comment]
@ambrad would you want to look at this again? I did not use a cmake function at the end. Standalone run was ok with these changes, cime run is in q.

**Comment by oksanaguba** [comment]
New tensor tests did not run properly here. We usually rely on baselines fail, but baselines failed, but -compare run proceeded. It is an easy fix for tensor nl. However there are issues: For baselines, Fortran run 'ran successfully' and only cxx run failed (while they both failed):
```
test preqx-nlev72-moist-r3-samenu-tensorhv-lim9-q6-ne2-ndays1-run was run successfully
Running test preqx-nlev72-moist-r3-samenu-tensorhv-lim9-q6-kokkos-ne2-ndays1-run ... /gpfs1/onguba/acme_scratch/sandiatoss3/HOMMEBFB_P24.f19_g16_rx1.A.sandiatoss3_intel.20191210_110720_t138lx/bld/tests/baseline/preqx-nlev72-moist-r3-samenu-tensorhv-lim9-q6-kokkos-ne2-ndays1/preqx-nlev72-moist-r3-samenu-tensorhv-lim9-q6-kokkos-ne2-ndays1-run.sh > preqx-nlev72-moist-r3-samenu-tensorhv-lim9-q6-kokkos-ne2-ndays1-run.out 2> preqx-nlev72-moist-r3-samenu-tensorhv-lim9-q6-kokkos-ne2-ndays1-run.err
failed with message:
cat: preqx-nlev72-moist-r3-samenu-tensorhv-lim9-q6-kokkos-ne2-ndays1-run.err: No such file or directory
```

Thus
1. Mechanism to detect which test ran ok and which did not is broken in homme (or broken in new xx tests).
2. If we run homme thru cime without -g or -c, baseline stage may fail but it won't show up as a FAIL. Status is reported for the last run, compare, which may pass because there could be 1 frame in failing tests.

I will fix tensor namelists and make sure when new tests run, they have 2 frames. Will insert another status check in homme.py .

**Comment by oksanaguba** [comment]
New tests were tweaked. By some reason lowers moist and tensor cases are more challenging. New namelists were run for all 3 tests for nightlies settings -- ne12days9 for const hv and ne14days3 for tensor hv (it is crashing for 9 day run). This current commit runs short profile, that is dev and short runs. all runs have 2 frames in movies, both short and dev. I checked F runs and cxx vs F tests all pass. Running this from cime now.

**Comment by oksanaguba** [comment]
Cime runs for HOMMEBFB and HOMME suites looks good. @jgfouca this is read, thanks!

**Comment by jgfouca** [comment]
pushed to next

**Pull request merged**